### PR TITLE
feat: mini workspace bubble

### LINKED ISSUE #7045: Refactor the "mutator bubble" to be a MiniWorkspaceBubble
### Description

Part of #2105 

Refactor the bubble UI used by mutator icons into its own class so that it can easily be reused by multiple icons.

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [X] I ran `npm run format` and `npm run lint`

### Resolves

Fixes #7045 

### Proposed Changes

Adds a new kind of bubble which is a mini workspace bubble.

### Reason for Changes

We want bubbles to be pure views with a object oriented structure. This works toward that =)

### Test Coverage

Manually tested by hacking into the warning icon. Looks the same as bubbles did previously.

This is the one I'm most nervous about since it has so much functionality that I can't test it all. But the simple case looks good! And I added validation to hopefully restrict peeps to those use cases.

If we want we can also mark the constructor `@internal` until we're more confident in it.

### Documentation

N/A

### CODE REVIEW

File: core/bubbles/mini_workspace_bubble.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    this.svgDialog = dom.createSvgElement(
+      Svg.SVG,
+      {
+        'x': Bubble.BORDER_WIDTH,
+        'y': Bubble.BORDER_WIDTH,
+      },
+      this.contentContainer
+    );
+    workspaceOptions.parentWorkspace = this.workspace;
+    this.miniWorkspace = this.newWorkspaceSvg(new Options(workspaceOptions));
```
Comment (cpcallen) [inline core/bubbles/mini_workspace_bubble.ts:42]:
```suggestion
    this.miniWorkspace = new WorkspaceSvg(new Options(workspaceOptions));
```

File: core/bubbles/mini_workspace_bubble.ts [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+  /** @internal */
+  getWorkspace(): WorkspaceSvg {
+    return this.miniWorkspace;
+  }
+
+  /** Adds a change listener to the mini workspace. */
+  addWorkspaceChangeListener(listener: (e: AbstractEvent) => void) {
+    this.miniWorkspace.addChangeListener(listener);
+  }
```
Comment (cpcallen) [inline core/bubbles/mini_workspace_bubble.ts:76]: What is the value of this method?  Why would a developer not just do `myMWB.getWorkspace().addChangeListener(…)`?  (Do we want to be able to monitor the adding of change listeners?  Seems futile given that the developer can just add it directly as noted.)

File: core/bubbles/mini_workspace_bubble.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+   * mini workspace.
+   */
+  private updateBubbleSize() {
+    if (this.miniWorkspace.isDragging()) return;
+
+    const currSize = this.getSize();
+    const newSize = this.calculateWorkspaceSize();
+    if (
+      Math.abs(currSize.width - newSize.width) < 10 &&
+      Math.abs(currSize.height - newSize.height) < 10
```
Comment (cpcallen) [inline core/bubbles/mini_workspace_bubble.ts:125]: Should this delta value (`10`) be configurable, or at least a named constant?  Also, what units is it measured in?  10px seems like a pretty noticeable change.

File: core/bubbles/mini_workspace_bubble.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      return;
+    }
+    this.svgDialog.setAttribute('width', `${newSize.width}px`);
+    this.svgDialog.setAttribute('height', `${newSize.height}px`);
+    this.miniWorkspace.setCachedParentSvgSize(newSize.width, newSize.height);
+    if (this.miniWorkspace.RTL) {
+      this.miniWorkspace
+        .getCanvas()
+        .setAttribute('transform', `translate(${newSize.width}, 0)`);
+    }
```
Comment (cpcallen) [inline core/bubbles/mini_workspace_bubble.ts:137]: I'm not sure what's going on here.  A comment would be helpful.  I guess this is because the workspace needs to have its left edge rather than right edge moved, but I'm unclear on why the code uses only the new width rather than the difference between new and old.

File: core/bubbles/mini_workspace_bubble.ts [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    this.miniWorkspace.recordDragTargets();
+  }
+
+  /**
+   * Calculates the size of the mini workspace for use in resizing the bubble.
+   */
+  private calculateWorkspaceSize(): Size {
+    const canvas = this.miniWorkspace.getCanvas();
+    const bbox = canvas.getBBox();
+    let width = this.miniWorkspace.RTL ? -bbox.x : bbox.width;
```
Comment (cpcallen) [inline core/bubbles/mini_workspace_bubble.ts:155]: Similarly to above: it eludes me why only `x` is relevant in RTL.  Surely a width is a width?

File: core/bubbles/mini_workspace_bubble.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    let height = bbox.height + bbox.y + Bubble.DOUBLE_BORDER * 3;
+    const flyout = this.miniWorkspace.getFlyout();
+    if (flyout) {
+      const flyoutScrollMetrics = flyout
+        .getWorkspace()
+        .getMetricsManager()
+        .getScrollMetrics();
```
Comment (cpcallen) [inline core/bubbles/mini_workspace_bubble.ts:162]: Maybe I am beginning to regret my previous endorsement of Prettier.

File: core/bubbles/mini_workspace_bubble.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+      height = Math.max(height, flyoutScrollMetrics.height + 20);
+      if (this.miniWorkspace.RTL) {
+        width = Math.max(width, flyout.getWidth());
+      } else {
+        width += flyout.getWidth();
+      }
+    }
+    width += Bubble.DOUBLE_BORDER * 3;
```
Comment (cpcallen) [inline core/bubbles/mini_workspace_bubble.ts:170]: Huh?  Why do we want to increase the width by 6 * border width?

File: core/bubbles/mini_workspace_bubble.ts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  /** @internal */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  newWorkspaceSvg(options: Options): WorkspaceSvg {
+    throw new Error(
+      'The implementation of newWorkspaceSvg should be ' +
+        'monkey-patched in by blockly.ts'
+    );
+  }
```
Comment (cpcallen) [inline core/bubbles/mini_workspace_bubble.ts:198]:
```suggestion
```

File: core/flyout_base.ts [hunk 9]
```diff
@@ -136,14 +137,14 @@ export abstract class Flyout extends DeleteArea implements IFlyout {
    * Function that will be registered as a change listener on the workspace
    * to reflow when blocks in the flyout workspace change.
    */
-  private reflowWrapper: Function | null = null;
+  private reflowWrapper: ((e: AbstractEvent) => void) | null = null;
```
Comment (cpcallen) [inline core/flyout_base.ts:140]: Hmm.  Technically this (and the below) is a breaking change, but I imagine that passing in a function that does not accept an `AbstractEvent` as its first parameter did not actually work previously, despite the sloppy signature.  Is that true?

File: core/mutator.ts [hunk 10]
```diff
@@ -265,7 +265,7 @@ export class Mutator extends Icon {
     const canvas = this.workspace_.getCanvas();
     const workspaceSize = canvas.getBBox();
     let width = workspaceSize.width + workspaceSize.x;
-    let height = workspaceSize.height + doubleBorderWidth * 3;
+    let height = workspaceSize.height + doubleBorderWidth * 2;
```
Comment (cpcallen) [inline core/mutator.ts:268]: Why this seemingly unrelated change to `Mutator`?

File: core/workspace.ts [hunk 11]
```diff
@@ -668,7 +668,7 @@ export class Workspace implements IASTNodeLocation {
    * @param func Function to call.
    * @returns Obsolete return value, ignore.
    */
-  addChangeListener(func: Function): Function {
+  addChangeListener(func: (e: Abstract) => void): Function {
```
Comment (cpcallen) [inline core/workspace.ts:671]: Ditto previous re: breaking change.

**Review by cpcallen — CHANGES_REQUESTED** [review]
> We want bubbles to be pure views with a object oriented structure. This works toward that =)

I'm not quite sure what you mean by "pure views".

Am I correct in understanding that the current `Mutator` (really: block button with gear icon) will in future use a `MiniWorkspaceBubble` instead of whatever it currently uses?

> This is the one I'm most nervous about since it has so much functionality that I can't test it all. But the simple case looks good! And I added validation to hopefully restrict peeps to those use cases.

I was surprised that the mini workspace doesn't support scrolling, trash, etc.  Are there technical limitations why that is, or is this just policy?  If the latter, why?

> If we want we can also mark the constructor `@internal` until we're more confident in it.

That seems like a good idea, though not exporting it (as is presently the case) is probably just as effective.

**Review by cpcallen — CHANGES_REQUESTED** [review]
> We want bubbles to be pure views with a object oriented structure. This works toward that =)

I'm not quite sure what you mean by "pure views".

Am I correct in understanding that the current `Mutator` (really: block button with gear icon) will in future use a `MiniWorkspaceBubble` instead of whatever it currently uses?

> This is the one I'm most nervous about since it has so much functionality that I can't test it all. But the simple case looks good! And I added validation to hopefully restrict peeps to those use cases.

I was surprised that the mini workspace doesn't support scrolling, trash, etc.  Are there technical limitations why that is, or is this just policy?  If the latter, why?

> If we want we can also mark the constructor `@internal` until we're more confident in it.

That seems like a good idea, though not exporting it (as is presently the case) is probably just as effective.

File: core/flyout_base.ts (hunk 9 above)
Comment (BeksOmega) [reply core/flyout_base.ts:140]: If you passed in a `() => void` that's fine because it's a subtype of `(anyParameters) => void`. If you passedin `(someTypeBesidesAnEvent) => void` then yeah that didn't work before. So this shouldn't break anything.

File: core/mutator.ts (hunk 10 above)
Comment (BeksOmega) [reply core/mutator.ts:268]: Accidental commit, thanks for the catch!

File: core/bubbles/bubble.ts [hunk 12]
```diff
@@ -20,8 +20,11 @@ export abstract class Bubble implements IBubble {
   /** The width of the border around the bubble. */
   static BORDER_WIDTH = 6;
 
+  /** Double the width of the border around the bubble. */
+  static DOUBLE_BORDER = this.BORDER_WIDTH * 2;
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply core/bubbles/bubble.ts:24]: Ah whoops, these should both be readonly.

File: core/bubbles/mini_workspace_bubble.ts (hunk 2 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts:76]: `getWorkspace` is internal because allowing them to arbitrarily mess with the workspace is a large promise that invites a lot of edge cases I don't have time to test. This creates a more minimal API.

File: core/bubbles/mini_workspace_bubble.ts (hunk 3 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts:125]: Can add a constant. I don't think the change is necessarily "noticable" because we always have margin around the contents of the workspace. So if we don't resize for <10px there will just be slightly less margin than if we do resize.

File: core/bubbles/mini_workspace_bubble.ts (hunk 4 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts:137]: I'll pull over the comment from `mutator`. I don't think the comment illuminates a lot, but yeah it does have to do with alignment.

File: core/bubbles/mini_workspace_bubble.ts (hunk 7 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts:170]: This is just an arbitrary margin, pulled over from the existing mutator code.

**Comment by BeksOmega** [comment]
> Am I correct in understanding that the current Mutator (really: block button with gear icon) will in future use a MiniWorkspaceBubble instead of whatever it currently uses?

Yes!

> I was surprised that the mini workspace doesn't support scrolling, trash, etc. Are there technical limitations why that is, or is this just policy? If the latter, why?

There are technical limitations related to how the workspace gets sized (and automatically resized). We can't support scrolling because the workspace should never be scrollable (always resizing to show the whole workspace). We can't support the trashcan because (currently) we don't take the size of the trashcan into account when resizing.

In the future we can always add support for these things! I just don't have time to implement that right now.

> That seems like a good idea, though not exporting it (as is presently the case) is probably just as effective.

I think we do want to export this for documentation purposes though. I'm pretty sure if it's not exported then API extractor won't include information about what other methods it has and what not. I'll mark it as internal just in case.

File: core/blockly.ts [hunk 13]
```diff
@@ -443,6 +444,12 @@ Mutator.prototype.newWorkspaceSvg = function (options: Options): WorkspaceSvg {
   return new WorkspaceSvg(options);
 };
 
+MiniWorkspaceBubble.prototype.newWorkspaceSvg = function (
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply core/blockly.ts:447]: Currently this is fine, but eventually we're going to have: WorkspaceSvg imports BlockSvg imports MutatorIcon imports MiniWorkspaceBubble imports WorkspaceSvg, and then it's going to break again. So might as well just add this from the beginning.

Also testing this with the WarningIcon caused the same problems.

Afaict this circular dependency is just an unfixable constraint if we want to provide miniworkspace bubbles :/

File: core/bubbles/mini_workspace_bubble.ts (hunk 5 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts:155]: Also eludes me, this was pulled over from the mutator code, but I don't think it was working 100% correctly. E.g. having a RTL bubble in a LTR workspace would break (idk why you would want to do that, but that's beside the point :P) Tested unifying the width code and it looks good! Renders correctly in LTR and RTL with a flyout and without a flyout.

File: core/blockly.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (cpcallen) [reply core/blockly.ts:447]: Hmm.  That is unfortunate.  I cant' help but think that there must be a better way to break this circularity.  Like: couldn't `MutatorIcon` just fulfil an `IIcon` interface, and have `Block` not know any more specifics?

But fixing that is probably out-of-scope for this PR, so for now I approve of you using the existing pattern.

File: core/bubbles/mini_workspace_bubble.ts (hunk 2 above)
Comment (cpcallen) [reply core/bubbles/mini_workspace_bubble.ts:91]: > `getWorkspace` is internal

Oh, good point.

I'm sort of tempted to suggest having `getWorkspace` be public but return a `Proxy` that strictly limits what kind of messing they can do, but I will refrain.  (You can tell that I have spent too much time thinking about adversarial programming in JavaScript for Code City.)

Question, though: if they can add a listener, can the listener obtain a reference to the `Workspace` when an event fires?  (See previous parenthetical comment.)

File: core/bubbles/mini_workspace_bubble.ts (hunk 3 above)
Comment (cpcallen) [reply core/bubbles/mini_workspace_bubble.ts]: Isn't the border only 6px?  Or am I confused?

File: core/bubbles/mini_workspace_bubble.ts (hunk 2 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts:91]: That's true, they can obtain a reference to the event via `Blockly.Workspace.getById` from the `workspaceId` property of the event.

File: core/bubbles/mini_workspace_bubble.ts (hunk 7 above)
Comment (cpcallen) [reply core/bubbles/mini_workspace_bubble.ts]: OK, I'm clearly confused between border and margin, but I think a comment (or a named constant) would be in order here too.

File: core/flyout_base.ts (hunk 9 above)
Comment (cpcallen) [reply core/flyout_base.ts:140]: Great.

File: core/bubbles/mini_workspace_bubble.ts (hunk 3 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts]: The border is 6px, but we also have white space that exists around the workspace, which is what I meant by margin (sorry we don't really have language for this stuff so I'm just making it up on the spot lol). 
image

**Comment by cpcallen** [comment]
> I think we do want to export this for documentation purposes though. I'm pretty sure if it's not exported then API extractor won't include information about what other methods it has and what not. I'll mark it as internal just in case.

Does API extractor include information about things marked `@internal` in its output?  And does it have any way to distinguish the type from the constructor for the type?  I'd sort of expect it would either show nothing about the type, or it would show the whole type as being `@internal`, which I think is not what you want.

But if you've checked that it is output as you intend then great.

**Review by cpcallen — APPROVED** [review]
Approved subject to due consideration given to remaining unresolved comments (changes at your discretion).

File: core/bubbles/mini_workspace_bubble.ts (hunk 5 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts]: I was testing this again to make sure I caught everything, and it doesn't resize correctly as we create space between the flyout and the block. I'm just going to pull over the exact code from the existing mutator (since we know that works in common situtations). And file a bug for actually fixing / explaining this logic, and fixing the erroneous bits (e.g. RTL in LTR mode).

File: core/bubbles/mini_workspace_bubble.ts (hunk 2 above)
Comment (BeksOmega) [reply core/bubbles/mini_workspace_bubble.ts:91]: I think that a proxy would be the most correct solution to this, but I think this is the best compromise for now. Neither the current solution or a proxy would provide complete protection due to `Blockly.Workspace.getById` addressed above (i.e. someone can still always get to the workspace if they want to). But this does limit external developer's access.

If I had more time I would investigate a proxy, but for now I'm going to consider it out of scope.

**Comment by BeksOmega** [comment]
> Does API extractor include information about things marked @internal in its output? ...

I will resolve this as part of #7082

File: core/bubbles/mini_workspace_bubble.ts (hunk 2 above)
Comment (cpcallen) [reply core/bubbles/mini_workspace_bubble.ts:97]: > If I had more time I would investigate a proxy, but for now I'm going to consider it out of scope.

The use of a `Proxy` was not really a serious suggestion (though it would certainly be a way to control access).  I think it would probably be better to invest the time in getting the flyout workspace to support all the regular features (including sub-bubbles!  🫧🫧🫧)
